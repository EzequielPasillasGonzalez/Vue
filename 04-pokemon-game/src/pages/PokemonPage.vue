<template>

  <h1 v-if="!pokemon">Espere por favor....</h1>

  <div v-else>
      <h1>¿Quién es este pokémon?</h1>
    
      <!-- todo: Img -->
      <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon"/> <!--? Se manda la properti de id al componente  -->    
      
      <!-- Todo: Opciones -->
      <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer"/> 
      <!--  ? Con el @selection, se escucha el evento que se creo en el componente hijo PokemonOPtions, 
            ? se puede llamar unicamente la erferencia al metodo, con los parentesis se puede mandar los argumentos 
            ? que se reciben, checkAnswer(1, $event) se pueden enviar tambien los que se generen ahi como el 1, y el $event es el que recibe 
            ? y manda a la funcion, tambien se puede dejar sin parentesis para pasar el unico elemento que recibe
      ?-->


      <template v-if="showAnswer"> <!--? Se peude usar div o template, funciona para agrupar cuando se van a usar las directivas -->
        <h2 class="fade-in">{{ message }}</h2>
        <button @click="newGame">Nuevo Juego</button>  
      </template>
      
  </div>
</template>

<script>
import PokemonOptions from '@/components/PokemonOptions.vue'
import PokemonPicture from '@/components/PokemonPicture.vue'
import getPokemonOptions from '@/helpers/getPokemonOptions'

export default {    
    components: {
    PokemonOptions,
    PokemonPicture
  },
  data(){
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: ''
    }
  },
  methods: {
    async mixPokemonArray(){
      this.pokemonArr = await getPokemonOptions()
      
      const rndInt = Math.floor(Math.random() * 4)
      this.pokemon = this.pokemonArr[rndInt]
    
    },

    checkAnswer(selectedIdPokemon){      
      this.showPokemon = true
      this.showAnswer = true

      if(selectedIdPokemon === this.pokemon.id){
        this.message = `Correct, ${this.pokemon.name}`
      }else{
        this.message = `False, ${this.pokemon.name}`
      }
    },

    newGame(){
      this.showPokemon = false
      this.showAnswer = false
      this.pokemonArr = []
      this.pokemon = null
      this.mixPokemonArray()      
    }
  },
  mounted(){
    this.mixPokemonArray()
  }
}
</script>

<style>

</style>