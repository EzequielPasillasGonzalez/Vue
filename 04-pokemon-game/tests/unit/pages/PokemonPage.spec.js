import { shallowMount } from '@vue/test-utils'
import PokemonPage from "@/pages/PokemonPage.vue";
import { mockPokemons } from '../mocks/pokemons.mock';

describe('PokemonPage Component', () => {

    let wrapper

    beforeEach(() => {
        wrapper = shallowMount(PokemonPage)
    })
    
    test('Debe de hacer match con el snapshot', () => { 

        expect(wrapper.html()).toMatchSnapshot()

    })

    test('Deeb de llamar el mixPokemonArray al montar', () => { 
        
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods, 'mixPokemonArray' ) //? Revisa los metodos u otra cosa
        wrapper = shallowMount(PokemonPage) //? Se debe ahcer asi cuando se quiere verificar algo en el mounted

        expect(mixPokemonArraySpy).toHaveBeenCalled()
    })

    test('Debe de hacer match con el snaptshot cuando cargan los pokemons', () => { 
        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        expect(wrapper.html()).toMatchSnapshot()
    })

    test('Debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        const pokemonPicture = wrapper.find('pokemon-picture-stub')
        const PokemonOptions = wrapper.find('pokemon-options-stub')

        //? PokemonPicture debe de existir
        expect(pokemonPicture).toBeTruthy()
        
        
        //? PokemonOptions debe de existir
        expect(PokemonOptions).toBeTruthy()

        //? Pokemon Picture attribute pokemonId === del arreglo en la posicion 0
        expect(pokemonPicture.attributes('pokemonid')).toEqual(mockPokemons[0].id.toString())

        //? PokemonOptions attribute pokemons toBe true
        expect(PokemonOptions.attributes('pokemons')).toBeTruthy()


    });

    test('Pruebas con checkAnswer', async () => {

        const wrapper = shallowMount(PokemonPage, {
            data() {
                return {
                    pokemonArr: mockPokemons,
                    pokemon: mockPokemons[0],
                    showPokemon: false,
                    showAnswer: false,
                    message: ''
                }
            }
        })

        await wrapper.vm.checkAnswer(0)
        
        expect(wrapper.find('h2').exists()).toBeTruthy()
        expect(wrapper.vm.showPokemon).toBe(true)
        expect(wrapper.find('h2').text()).toBe(`False, ${mockPokemons[0].name}`)
        
        
        
    })

});