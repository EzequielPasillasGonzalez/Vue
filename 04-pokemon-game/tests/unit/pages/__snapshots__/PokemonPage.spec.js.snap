// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component Debe de hacer match con el snapshot 1`] = `<h1>Espere por favor....</h1>`;

exports[`PokemonPage Component Debe de hacer match con el snaptshot cuando cargan los pokemons 1`] = `
<div>
  <h1>¿Quién es este pokémon?</h1><!-- todo: Img -->
  <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
  <!--? Se manda la properti de id al componente  -->
  <!-- Todo: Opciones -->
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--  ? Con el @selection, se escucha el evento que se creo en el componente hijo PokemonOPtions, 
            ? se puede llamar unicamente la erferencia al metodo, con los parentesis se puede mandar los argumentos 
            ? que se reciben, checkAnswer(1, $event) se pueden enviar tambien los que se generen ahi como el 1, y el $event es el que recibe 
            ? y manda a la funcion, tambien se puede dejar sin parentesis para pasar el unico elemento que recibe
      ?-->
  <!--v-if-->
</div>
`;
